halo/app/ready.go

```diff
@@ -13,39 +13,40 @@ type readinessStatus struct {
 	mu                 sync.RWMutex
 	ConsensusSynced    bool    `json:"consensus_synced"`              // Unhealthy if false
 	ConsensusP2PPeers  int     `json:"consensus_p2p_peers"`           // Unhealthy if 0
+	ConsensusHeight    int64   `json:"consensus_height"`              // Unhealthy if 0
+	ConsensusRunning   bool    `json:"consensus_running"`             // Unhealthy if false
 	ExecutionConnected bool    `json:"execution_connected"`           // Unhealthy if false
 	ExecutionSynced    bool    `json:"execution_synced"`              // Unhealthy if false
-	ExecutionP2PPeers  *uint64 `json:"execution_p2p_peers,omitempty"` // Unhealthy if 0
+	ExecutionP2PPeers  *uint64 `json:"execution_p2p_peers,omitempty"` // Unhealthy if 0 (ignored if not available)
+	ExecutionHeight    uint64  `json:"execution_height"`              // Unhealthy if 0
 }
 
 // ready indicates if the current status is considered as healthy.
-// If the current halo instance is using an EVM RPC endpoint,
-// the peers count is not taken into the account.
 func (r *readinessStatus) ready() bool {
 	r.mu.RLock()
 	defer r.mu.RUnlock()
 
-	ready := r.ConsensusSynced && r.ConsensusP2PPeers > 0 && r.ExecutionConnected &&
-		r.ExecutionSynced
-
-	peers := r.ExecutionP2PPeers
-	if peers != nil {
-		return ready && *peers > 0
-	}
-
-	return ready
+	return r.ConsensusSynced &&
+		r.ConsensusP2PPeers > 0 &&
+		r.ConsensusHeight > 0 &&
+		r.ConsensusRunning &&
+		r.ExecutionConnected &&
+		r.ExecutionSynced &&
+		r.ExecutionHeight > 0 &&
+		(r.ExecutionP2PPeers == nil || *r.ExecutionP2PPeers > 0)
 }
 
 // serialize marshals the struct to the specified writer.
-func (r *readinessStatus) serialize(w io.Writer) error {
+// It also returns the ready status.
+func (r *readinessStatus) serialize(w io.Writer) (bool, error) {
 	r.mu.RLock()
 	defer r.mu.RUnlock()
 
 	if err := json.NewEncoder(w).Encode(r); err != nil {
-		return errors.Wrap(err, "readiness status serialization")
+		return false, errors.Wrap(err, "readiness status serialization")
 	}
 
-	return nil
+	return r.ready(), nil
```

Comment:
Wouldn't this cause a dead lock? `ready()` will try to lock while we're holding the lock?

Reply:
Ah looks like read locks are reentrant in Go by default. This explains why no test is failing.